elasticapm/instrumentation/packages/urllib.py

```diff
@@ -0,0 +1,102 @@
+#  modification, are permitted provided that the following conditions are met:
+#
+#  * Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+#  * Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+#  * Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+#  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+#  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+#  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+#  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+#  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+from elasticapm.conf import constants
+from elasticapm.instrumentation.packages.base import AbstractInstrumentedModule
+from elasticapm.traces import DroppedSpan, capture_span, execution_context
+from elasticapm.utils import compat, default_ports
+from elasticapm.utils.disttracing import TracingOptions
+
+
+# copied and adapted from urllib.request
+def request_host(request):
+    """Return request-host, as defined by RFC 2965.
+
+    Variation from RFC: returned value is lowercased, for convenient
+    comparison.
+
+    """
+    url = request.get_full_url()
+    scheme, host = compat.urlparse.urlparse(url)[:2]
+    if host == "":
+        host = request.get_header("Host", "")
+    if ":" in host:
+        bare_host, port = host.split(":")
+        try:
+            port = int(port)
+        except ValueError:
+            pass
+        if port == default_ports.get(scheme):
+            return bare_host
+    return host
+
+
+class UrllibInstrumentation(AbstractInstrumentedModule):
+    name = "urllib"
+
+    if compat.PY2:
+        instrument_list = [("urllib2", "AbstractHTTPHandler.do_open")]
+    else:
+        instrument_list = [("urllib.request", "AbstractHTTPHandler.do_open")]
+
+    def call(self, module, method, wrapped, instance, args, kwargs):
+        request_object = args[1] if len(args) > 1 else kwargs["req"]
+
+        method = request_object.get_method()
+        host = request_host(request_object)
+
+        url = request_object.get_full_url()
```

Comment:
Is it possible for this to include a user:password? as in "http://user:pass@foo.bar/baz". It would be good to see a test that ensures we don't leak application secrets.

Reply:
urllib (as well as urllib3) doesn't support encoding user/password in the URL, so this would only be an issue for code that is buggy in the first place.

`requests` OTOH does support this according to my testing (even though it isn't documented in https://2.python-requests.org//en/master/user/authentication/#basic-authentication), so we need the sanitization code anyway.

Reply:
done in e8826da